File: entities.py
"""
Game entities: Player, Cookies, and Chutes
"""

import pygame
import math
from constants import *


class Player:
    """The Fean character that catches cookies"""
    
    def __init__(self, x, y):
        """Initialize the player"""
        self.x = x
        self.y = y
        self.width = PLAYER_WIDTH
        self.height = PLAYER_HEIGHT
        self.velocity_x = 0
        self.speed = PLAYER_SPEED
    
    def move_left(self):
        """Move player left"""
        self.velocity_x = -self.speed
    
    def move_right(self):
        """Move player right"""
        self.velocity_x = self.speed
    
    def update(self):
        """Update player position"""
        self.x += self.velocity_x
        
        # Boundary checking
        if self.x - self.width // 2 < 0:
            self.x = self.width // 2
        elif self.x + self.width // 2 > SCREEN_WIDTH:
            self.x = SCREEN_WIDTH - self.width // 2
        
        # Friction/deceleration
        self.velocity_x *= 0.9
    
    def draw(self, surface):
        """Draw the player"""
        # Draw a simple rectangle for now (will be replaced with sprite)
        player_rect = pygame.Rect(self.x - self.width // 2,
                                   self.y - self.height // 2,
                                   self.width,
                                   self.height)
        
        # Draw player body (blue to match FEAN shirt)
        pygame.draw.rect(surface, PLAYER_COLOR, player_rect, border_radius=10)
        
        # Draw a simple face
        face_center_x = self.x
        face_center_y = self.y - self.height // 3
        pygame.draw.circle(surface, SKIN_COLOR, (face_center_x, face_center_y), self.width // 6)
        
        # Draw eyes
        eye_offset = self.width // 10
        pygame.draw.circle(surface, BLACK, (face_center_x - eye_offset, face_center_y - self.width // 15), 3)
        pygame.draw.circle(surface, BLACK, (face_center_x + eye_offset, face_center_y - self.width // 15), 3)
        
        # Draw a catching basket area (lighter color)
        basket_rect = pygame.Rect(self.x - self.width // 3,
                                   self.y + self.height // 4,
                                   self.width // 1.5,
                                   self.height // 3)
        pygame.draw.rect(surface, BASKET_COLOR, basket_rect, border_radius=5)


class Cookie:
    """A cookie falling from the chutes"""
    
    def __init__(self, x, y):
        """Initialize a cookie"""
        self.x = x
        self.y = y
        self.velocity_y = COOKIE_FALL_SPEED
        self.size = COOKIE_SIZE
        self.rotation = 0
        self.rotation_speed = 5
    
    def update(self):
        """Update cookie position and rotation"""
        self.y += self.velocity_y
        self.velocity_y += GRAVITY  # Accelerate downward
        self.rotation += self.rotation_speed
        
        # Cap maximum velocity
        if self.velocity_y > MAX_COOKIE_VELOCITY:
            self.velocity_y = MAX_COOKIE_VELOCITY
    
    def draw(self, surface):
        """Draw the cookie"""
        # Draw cookie circle
        pygame.draw.circle(surface, COOKIE_COLOR, (int(self.x), int(self.y)), self.size // 2)
        
        # Draw cookie chips (small dots)
        chip_count = 8
        for i in range(chip_count):
            angle = (360 / chip_count) * i + (self.rotation % 360)
            rad = math.radians(angle)
            chip_x = self.x + math.cos(rad) * (self.size // 3)
            chip_y = self.y + math.sin(rad) * (self.size // 3)
            pygame.draw.circle(surface, CHIP_COLOR, (int(chip_x), int(chip_y)), 2)


class Chute:
    """A cookie chute that dispenses cookies"""
    
    def __init__(self, x, y):
        """Initialize a chute"""
        self.x = x
        self.y = y
        self.width = CHUTE_WIDTH
        self.height = CHUTE_HEIGHT
    
    def draw(self, surface):
        """Draw the chute"""
        # Draw chute body
        chute_rect = pygame.Rect(self.x - self.width // 2,
                                  self.y,
                                  self.width,
                                  self.height)
        pygame.draw.rect(surface, CHUTE_COLOR, chute_rect, border_radius=5)
        
        # Draw chute opening (darker)
        opening_rect = pygame.Rect(self.x - self.width // 3,
                                    self.y + self.height - 10,
                                    self.width // 1.5,
                                    10)
        pygame.draw.rect(surface, CHUTE_OPENING_COLOR, opening_rect)
        
        # Draw chute outline
        pygame.draw.rect(surface, BLACK, chute_rect, 2, border_radius=5)


class Particle:
    """A particle effect (for catching cookies, etc.)"""
    
    def __init__(self, x, y, vx, vy, lifetime=30, color=(255, 215, 0)):
        """Initialize a particle"""
        self.x = x
        self.y = y
        self.vx = vx
        self.vy = vy
        self.lifetime = lifetime
        self.max_lifetime = lifetime
        self.color = color
        self.size = 5
    
    def update(self):
        """Update particle position"""
        self.x += self.vx
        self.y += self.vy
        self.vy += GRAVITY
        self.lifetime -= 1
    
    def draw(self, surface):
        """Draw the particle"""
        # Fade out over time
        alpha = int(255 * (self.lifetime / self.max_lifetime))
        color = tuple(min(255, c) for c in self.color)
        
        pygame.draw.circle(surface, color, (int(self.x), int(self.y)), self.size)
    
    def is_alive(self):
        """Check if particle is still alive"""
        return self.lifetime > 0

File: js/entities.js
// Game Entities

class Player {
    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.width = PLAYER_WIDTH;
        this.height = PLAYER_HEIGHT;
        this.velocityX = 0;
        this.speed = PLAYER_SPEED;
    }

    moveLeft() {
        this.velocityX = -this.speed;
    }

    moveRight() {
        this.velocityX = this.speed;
    }

    update() {
        this.x += this.velocityX;

        // Boundary checking
        if (this.x - this.width / 2 < 0) {
            this.x = this.width / 2;
        } else if (this.x + this.width / 2 > SCREEN_WIDTH) {
            this.x = SCREEN_WIDTH - this.width / 2;
        }

        // Friction
        this.velocityX *= 0.9;
    }

    draw(ctx) {
        // Draw player body (rectangle)
        ctx.fillStyle = COLORS.PLAYER;
        ctx.fillRect(
            this.x - this.width / 2,
            this.y - this.height / 2,
            this.width,
            this.height
        );

        // Draw rounded corners effect
        ctx.strokeStyle = COLORS.DARK_BLUE;
        ctx.lineWidth = 3;
        ctx.strokeRect(
            this.x - this.width / 2,
            this.y - this.height / 2,
            this.width,
            this.height
        );

        // Draw player head (circle)
        const headCenterX = this.x;
        const headCenterY = this.y - this.height / 3;
        const headRadius = this.width / 6;

        ctx.fillStyle = COLORS.SKIN;
        ctx.beginPath();
        ctx.arc(headCenterX, headCenterY, headRadius, 0, Math.PI * 2);
        ctx.fill();

        // Draw eyes
        const eyeOffset = this.width / 10;
        const eyeY = headCenterY - this.width / 15;
        
        ctx.fillStyle = COLORS.BLACK;
        ctx.beginPath();
        ctx.arc(headCenterX - eyeOffset, eyeY, 4, 0, Math.PI * 2);
        ctx.fill();

        ctx.beginPath();
        ctx.arc(headCenterX + eyeOffset, eyeY, 4, 0, Math.PI * 2);
        ctx.fill();

        // Draw smile
        ctx.strokeStyle = COLORS.BLACK;
        ctx.lineWidth = 2;
        ctx.beginPath();
        ctx.arc(headCenterX, headCenterY + 5, 8, 0, Math.PI);
        ctx.stroke();

        // Draw basket (catching area)
        ctx.fillStyle = COLORS.BASKET;
        ctx.fillRect(
            this.x - this.width / 3,
            this.y + this.height / 4,
            this.width / 1.5,
            this.height / 3
        );

        ctx.strokeStyle = COLORS.DARK_BLUE;
        ctx.lineWidth = 2;
        ctx.strokeRect(
            this.x - this.width / 3,
            this.y + this.height / 4,
            this.width / 1.5,
            this.height / 3
        );
    }

    getCollisionRect() {
        return {
            x: this.x - this.width / 2,
            y: this.y - this.height / 2,
            width: this.width,
            height: this.height
        };
    }
}

class Cookie {
    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.velocityY = COOKIE_FALL_SPEED;
        this.size = COOKIE_SIZE;
        this.rotation = 0;
        this.rotationSpeed = 5;
    }

    update() {
        this.y += this.velocityY;
        this.velocityY += GRAVITY;

        // Cap maximum velocity
        if (this.velocityY > MAX_COOKIE_VELOCITY) {
            this.velocityY = MAX_COOKIE_VELOCITY;
        }

        this.rotation += this.rotationSpeed;
    }

    draw(ctx) {
        // Draw cookie circle
        ctx.fillStyle = COLORS.COOKIE;
        ctx.beginPath();
        ctx.arc(this.x, this.y, this.size / 2, 0, Math.PI * 2);
        ctx.fill();

        // Draw border
        ctx.strokeStyle = COLORS.CHIP;
        ctx.lineWidth = 2;
        ctx.stroke();

        // Draw chocolate chips
        const chipCount = 8;
        ctx.fillStyle = COLORS.CHIP;

        for (let i = 0; i < chipCount; i++) {
            const angle = (360 / chipCount) * i + (this.rotation % 360);
            const rad = (angle * Math.PI) / 180;
            const chipX = this.x + Math.cos(rad) * (this.size / 3);
            const chipY = this.y + Math.sin(rad) * (this.size / 3);

            ctx.beginPath();
            ctx.arc(chipX, chipY, 3, 0, Math.PI * 2);
            ctx.fill();
        }
    }

    getCollisionRect() {
        return {
            x: this.x - this.size / 2,
            y: this.y - this.size / 2,
            width: this.size,
            height: this.size
        };
    }
}

class Chute {
    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.width = CHUTE_WIDTH;
        this.height = CHUTE_HEIGHT;
    }

    draw(ctx) {
        // Draw chute body
        ctx.fillStyle = COLORS.CHUTE;
        ctx.fillRect(
            this.x - this.width / 2,
            this.y,
            this.width,
            this.height
        );

        // Draw chute border
        ctx.strokeStyle = COLORS.BLACK;
        ctx.lineWidth = 2;
        ctx.strokeRect(
            this.x - this.width / 2,
            this.y,
            this.width,
            this.height
        );

        // Draw chute opening (darker)
        ctx.fillStyle = COLORS.CHUTE_OPENING;
        ctx.fillRect(
            this.x - this.width / 3,
            this.y + this.height - 10,
            this.width / 1.5,
            10
        );

        // Draw opening border
        ctx.strokeStyle = COLORS.BLACK;
        ctx.lineWidth = 1;
        ctx.strokeRect(
            this.x - this.width / 3,
            this.y + this.height - 10,
            this.width / 1.5,
            10
        );
    }
}

class Particle {
    constructor(x, y, vx, vy, lifetime = 30, color = '#ffd700') {
        this.x = x;
        this.y = y;
        this.vx = vx;
        this.vy = vy;
        this.lifetime = lifetime;
        this.maxLifetime = lifetime;
        this.color = color;
        this.size = 5;
    }

    update() {
        this.x += this.vx;
        this.y += this.vy;
        this.vy += GRAVITY;
        this.lifetime--;
    }

    draw(ctx) {
        const alpha = this.lifetime / this.maxLifetime;
        ctx.globalAlpha = alpha;

        ctx.fillStyle = this.color;
        ctx.beginPath();
        ctx.arc(this.x, this.y, this.size, 0, Math.PI * 2);
        ctx.fill();

        ctx.globalAlpha = 1.0;
    }

    isAlive() {
        return this.lifetime > 0;
    }
}

// Collision detection utility
function checkCollision(rect1, rect2) {
    return (
        rect1.x < rect2.x + rect2.width &&
        rect1.x + rect1.width > rect2.x &&
        rect1.y < rect2.y + rect2.height &&
        rect1.y + rect1.height > rect2.y
    );
}
